src/fedora_easyfix/utilities/producer.py

```diff
@@ -30,71 +30,93 @@
-                    gitlab_api_key
-                ).return_repository_collection()
-            ticket_collection["collection_updated_at"] = time()
-            write_index_to_local_json(ticket_collection)
+    def check_repolist_version_and_start(self):
+        if self.yamldict["repolist_version"] == __version__:
+            self.populate_ticket_collection()
+            self.ticket_collection["collection_updated_at"] = time()
+            self.write_index_to_local_json()
         else:
             statdcrt.failure("Could not index tickets")
             statdcrt.general("Repolist version does not correspond with the Easyfix version")
+            exit()
+
+    def populate_ticket_collection(self):
+        statdcrt.section("Indexing tickets...")
+        if "github" in self.yamldict["forges"].keys():
+            github_repository_list = self.yamldict["forges"]["github"]["repositories"]
+            github_base_url = self.yamldict["forges"]["github"]["url"]
+            statdcrt.warning(
+                "Found %s repositories on GitHub" %
+                len(self.yamldict["forges"]["github"]["repositories"].keys())
+            )
+            self.ticket_collection["github"] = GitHubRepositories(
+                github_repository_list,
+                github_base_url,
+                self.github_api_key,
+                self.github_username
+            ).return_repository_collection()
+        if "pagure" in self.yamldict["forges"].keys():
+            pagure_repository_list = self.yamldict["forges"]["pagure"]["repositories"]
+            pagure_base_url = self.yamldict["forges"]["pagure"]["url"]
+            statdcrt.warning(
+                "Found %s repositories on Pagure" %
+                len(self.yamldict["forges"]["pagure"]["repositories"].keys())
+            )
+            self.ticket_collection["pagure"] = PagureRepositories(
+                pagure_repository_list,
+                pagure_base_url,
+                self.pagure_api_key
+            ).return_repository_collection()
+        if "gitlab" in self.yamldict["forges"].keys():
+            gitlab_repository_list = self.yamldict["forges"]["gitlab"]["repositories"]
+            gitlab_base_url = self.yamldict["forges"]["gitlab"]["url"]
+            statdcrt.warning(
+                "Found %s repositories on GitLab" %
+                len(self.yamldict["forges"]["gitlab"]["repositories"].keys())
+            )
+            self.ticket_collection["gitlab"] = GitLabRepositories(
+                gitlab_repository_list,
+                gitlab_base_url,
+                self.gitlab_api_key
+            ).return_repository_collection()
+
+    def write_index_to_local_json(self):
+        try:
+            tickdata = dumps(self.ticket_collection, indent=4)
+            with open("tickdata.json", "w") as tickfile:
+                tickfile.write(tickdata)
+            statdcrt.section("Indexing complete!")
+        except Exception as expt:
```

Comment:
I think catching such a wide range of exception is not a best practice, as this would (if I am not wrong) also catch SyntaxError (at least, it as the case in Python 2).

Also, it might be more resilient to write a temporary file, and then rename it. If the filesystem is full, I wonder if this wouldn't result in a empty file, and so no easyfix index. 

With a write + rename, if the write fail , the old index is still up, so less disruption.

But I am not 100% sure this is worth adding complexity, since we have monitoring, and we would spot that easyfix is empty quite fast.

Reply:
> I think catching such a wide range of exception is not a best practice, as this would (if I am not wrong) also catch SyntaxError (at least, it as the case in Python 2).

Valid point. The reason why I have kept a single clause to catch a wide range of exceptions is to keep the code resilient to most (if not all) runtime issues when deployed. Developers can remove them when they are working on the code and add it back later when they are done checking. I have assigned the exception to an `expt` variable as well for it to be used in debugging wherever needed.

> Also, it might be more resilient to write a temporary file, and then rename it. If the filesystem is full, I wonder if this wouldn't result in a empty file, and so no easyfix index.
> 
> With a write + rename, if the write fail , the old index is still up, so less disruption.
> 
> But I am not 100% sure this is worth adding complexity, since we have monitoring, and we would spot that easyfix is empty quite fast.

I do not think that we would have that less of space on the servers, owing to the fact that an index would hardly be taking ~150kB space on the disk. I think we are good for now but if needed, we can definitely introduce that change in future.